src/details/ArborX_DetailsConcepts.hpp

```diff
@@ -60,6 +61,91 @@ struct has_centroid<
 {
 };
 
+// is_complete implementation taken from https://stackoverflow.com/a/44229779
+template <class T, std::size_t = sizeof(T)>
+std::true_type is_complete_impl(T *);
+
+std::false_type is_complete_impl(...);
+
+template <class T>
+using is_complete = decltype(is_complete_impl(std::declval<T *>()));
+
+template <typename T, typename TTag, typename = void>
+struct has_access_traits : std::false_type
+{
+};
+
+template <typename T, typename TTag>
+struct has_access_traits<
+    T, TTag, std::enable_if_t<is_complete<Traits::Access<T, TTag>>::value>>
+    : std::true_type
+{
+};
+
+template <typename T, typename = void>
+struct has_memory_space : std::false_type
+{
+};
+
+template <typename T>
+struct has_memory_space<
+    T,
+    std::enable_if_t<Kokkos::is_memory_space<typename T::memory_space>::value>>
+    : std::true_type
+{
+};
+
+template <typename T>
+struct first_template_parameter;
```

Comment:
Could you have done this with `std::tuple` to extract the type at position 0?

Reply:
Yes